### PR TITLE
Move ZTP flow code to assisted-service

### PR DESCRIPTION
**Pull request by osherdp**
Migrate code from ``openshift/release`` to assisted-service.

### CODE REVIEW

File: deploy/operator/ztp/deploy_sno.sh [hunk 1]
```diff
@@ -0,0 +1,89 @@
+export ASSISTED_NAMESPACE="${ASSISTED_NAMESPACE:-assisted-installer}"
```
Comment (osherdp) [inline deploy/operator/ztp/deploy_sno.sh:1]: should we rename this file to ``deploy_spoke_cluster.sh``?

Comment (YuviGold) [reply deploy/operator/ztp/deploy_sno.sh:1]: yes

File: deploy/operator/ztp/deploy_sno.sh [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+export OPENSHIFT_INSTALL_RELEASE_IMAGE="${OPENSHIFT_INSTALL_RELEASE_IMAGE:-quay.io/openshift-release-dev/ocp-release:4.8.0-fc.3-x86_64}"
+export ASSISTED_CLUSTER_DEPLOYMENT_NAME="${ASSISTED_CLUSTER_DEPLOYMENT_NAME:-assisted-test-cluster}"
+export ASSISTED_AGENT_CLUSTER_INSTALL_NAME="${ASSISTED_AGENT_CLUSTER_INSTALL_NAME:-assisted-agent-cluster-install}"
+export ASSISTED_INFRAENV_NAME="${ASSISTED_INFRAENV_NAME:-assisted-infra-env}"
+export ASSISTED_PULLSECRET_NAME="${ASSISTED_PULLSECRET_NAME:-assisted-pull-secret}"
+export ASSISTED_PULLSECRET_JSON="${ASSISTED_PULLSECRET_JSON:-/home/test/dev-scripts/pull_secret.json}"
+export ASSISTED_PRIVATEKEY_NAME="${ASSISTED_PRIVATEKEY_NAME:-assisted-ssh-private-key}"
+export EXTRA_BAREMETALHOSTS_FILE="${EXTRA_BAREMETALHOSTS_FILE:-/home/test/dev-scripts/ocp/ostest/extra_baremetalhosts.json}"
+# TODO: make SSH public key configurable
+
```
Comment (YuviGold) [inline deploy/operator/ztp/deploy_sno.sh:13]: set -o nounset
set -o pipefail
set -o errexit
set -o xtrace

File: deploy/operator/ztp/deploy_sno.sh [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+oc create -f generated/clusterDeployment.yaml
+
+echo "Creating Assisted Installer InfraEnv"
+oc create -f generated/infraEnv.yaml
+
+echo "Creating Assisted Installer AgentClusterInstall"
+oc create -f generated/agentClusterInstall.yaml
+
+echo "Creating new baremetal host using baremetal operator"
+oc create -f generate/baremetalHost.yaml
```
Comment (YuviGold) [inline deploy/operator/ztp/deploy_sno.sh:59]: I don't think the `echo`s are needed. The `oc create` is verbose and outputs what it created

File: deploy/operator/ztp/deploy_sno.sh [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  fi
+done
+
+if [ "$status" != "running" ]
+then
+  echo "$vm never started"
+  exit 1
+else
+  echo "$vm was started"
+fi
```
Comment (YuviGold) [inline deploy/operator/ztp/deploy_sno.sh:89]: We don't really care about the VM. we should check if an `agent` CR was created matching the amount of hosts

File: deploy/operator/ztp/infraEnv.j2 [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+metadata:
+  name: '{{ infraenv_name }}'
+  namespace: '{{ assisted_namespace }}'
+spec:
+  clusterRef:
+    name: '{{ cluster_deployment_name }}'
+    namespace: '{{ assisted_namespace }}'
+  agentLabelSelector:
+    matchLabels:
+      bla: aaa
```
[the remark this answers is not in the document]
Comment (osherdp) [reply deploy/operator/ztp/infraEnv.j2:12]: I'll remove it

File: deploy/operator/ztp/clusterDeployment.j2 [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      agent:
+        sshPublicKey: '{{ ssh_public_key }}'
+        networking:
+          clusterNetwork:
+            - cidr: 10.128.0.0/14
+              hostPrefix: 23
+          serviceNetwork:
+            - 172.30.0.0/16
+        provisionRequirements:
+          controlPlaneAgents: 1
```
[the remark this answers is not in the document]
Comment (osherdp) [reply deploy/operator/ztp/clusterDeployment.j2:37]: I'll remove

File: deploy/operator/ztp/deploy_sno.sh (hunk 4 above)
Comment (osherdp) [reply deploy/operator/ztp/deploy_sno.sh:89]: right
fixed

**Review by YuviGold — APPROVED** [review]

**Comment by YuviGold** [comment]
@osherdp Does the ZTP job runs this code?

**Comment by osherdp** [comment]
@YuviGold not yet
after pushing it I will update the job
you can see how it will approximately look on https://github.com/openshift/release/pull/18826

**Pull request merged**